### DISCUSSION TITLE
Detect when player jumping

### DISCUSSION THREAD

**Discussion by Braayy** [Minestom/Minestom#1106] [Help]
Is there a way of detecting when a player jumps? I know you can get if the y changed in the PlayerMoveEvent, but other things can cause a positive y change so there is a better solution for it?

**Comment by ZakShearman** [comment]
IIRC if you're on stairs, a ladder, etc.. you're still classed as being on ground, although I'd have to double check. If so, the best you could do is check their Y velocity and that they're on ground.

That's all I can think of.

**Reply by ZakShearman** [reply]
Ok being on a ladder does not class as being on ground, so you'd also need to differentiate between being on a ladder.

**Comment by VendoAU** [comment]
You can compare a player's y motion like this:
```java
final double jumpMotion = 0.42F;

MinecraftServer.getGlobalEventHandler().addListener(PlayerMoveEvent.class, event -> {
    final Player player = event.getPlayer();
    final double motionY = event.getNewPosition().y() - player.getPosition().y();
    if (motionY == jumpMotion) {
        player.sendMessage("JUMPED");
    }
});
```
`jumpMotion` has to be a float which is then cast/stored as a double as that is how the client sends it.

This won't work when there is a block directly above the player.
There are some other cases that you may have to sort out, but this will work for most situations.

**Comment by VegetalDev** [comment]
Don't know if it's of any use, but I came up with a packet-based solution. Unfortunately, there is no actual "jump" packet, only one for player position:

```java
MinecraftServer.getPacketListenerManager().setListener(ClientPlayerPositionPacket.class, (packet, player) -> {
    Point newPoint = packet.position();
    Point lastPoint = player.getPosition();

    //players flying up probably shouldn't be considered constantly jumping
    if(player.isOnGround()  && !packet.onGround() && !player.isFlying() && newPoint.sub(lastPoint).y() > 0)
        System.out.println(player.getUsername() + " jumped!"); //custom event could go here instead

    //call Minestom's method to handle this packet after, not before, otherwise the player's position will be 
    //updated and we won't know the old position
    PlayerPositionListener.playerPositionListener(packet, player);
});
```

Since this listens to client packets directly, and is handled before physics is applied, it works even if there's a block above stopping the movement. However, players who start climbing a ladder will be considered "jumping", as will any other client-side actions that initiate a vertical change which moves the player upwards some amount. 

Walking up stairs and slabs does *not* count as a jump with this code; thankfully we can check `packet.onGround()`, which is `false` when the player is jumping but `true` when they try to step up a partial block. 

If you need to rule out server-initiated movement (stop it from triggering jump detection), I think you can add an additional check on the player's initial server-reported y-velocity. This may be a bit tricky though, because if I recall correctly, entities have some (possibly undefined?) amount of negative y-velocity even when not moving, due to how gravity works (related: #1107).

I would try to add a workaround for this, but my current testing environment is running a fork with a rewritten entity velocity system which doesn't handle gravity quite the same way. 

EDIT: you can eliminate the `isFlying` check as it is redundant considering we already check `isOnGround`, and players initiating flight are considered "jumping" either way. 

EDIT 2: you will need to handle `ClientPlayerPositionAndRotationPacket` in the same way, otherwise jumping will not be detected if the player is also rotating.

**Reply by SergeyLord** [reply]
I noticed a pattern:
length= 0.41999998688697815
...

And it's good for ladders.
but unfortunately climbing the stairs (and any slabs) is also considered a jump.

MinecraftServer.getGlobalEventHandler().addListener(PlayerMoveEvent.class, event -> {

            Player player = event.getPlayer();
            Pos pos1 = player.getPosition();
            Pos pos2 = event.getNewPosition();

            Vec dVec = pos2.sub(pos1).asVec();
            double length = dVec.y();

            if (Math.abs(length - 0.42) < 0.0001) {
                System.out.println("JUMP");
            }
     });
